streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java

```diff
@@ -106,8 +113,10 @@ public void onCompletion(final RecordMetadata metadata, final Exception exceptio
                             }
                             final TopicPartition tp = new TopicPartition(metadata.topic(), metadata.partition());
                             offsets.put(tp, metadata.offset());
-                        } else {
-                            if (sendException == null) {
+                        } else if (sendException == null) {
+                            ProductionExceptionHandlerResponse exceptionAction = productionExceptionHandler.handle(serializedRecord, exception);
```

Comment:
Why do we call the handler only for the first send exception? Should we do the callback for each?

Nit: add `final` to `ProductionExceptionHandlerResponse exceptionAction`

Also, for `ProducerFencedException` we are doomed to fail for any future record, too. Should we call the handler for this case? Note, `ProducerFencedException` will result in a self-healing behavior and Streams will not die. This raises the question, if we should distinguish between fatal exceptions that will occur for all records (eg, authorization) and record specific errors (like record too large) and only call the handler it if makes sense to skip over a record?

Reply:
> Why do we call the handler only for the first send exception? Should we do the callback for each?

My thinking here was that once the decision had been made to `FAIL`, any further invocations of the handler would just be extra noise.

> Nit: add final to ProductionExceptionHandlerResponse exceptionAction

Sure

> Also, for ProducerFencedException we are doomed to fail for any future record, too. Should we call the handler for this case? Note, ProducerFencedException will result in a self-healing behavior and Streams will not die. This raises the question, if we should distinguish between fatal exceptions that will occur for all records (eg, authorization) and record specific errors (like record too large) and only call the handler it if makes sense to skip over a record?

I responded to this on the mailing list a bit, but you didn't mention `ProducerFencedException`.

I'm inclined to think that we should not invoke the handler for `ProducerFencedException` if it is always self-healing, but I did like the idea of giving the users hooks to cause their own side effects....

I need to think on this, but I'm really not thrilled with how this feels at the moment.